presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PayloadJoinOptimizer.java

```diff
@@ -215,8 +221,13 @@ public PlanNode visitJoin(JoinNode joinNode, RewriteContext<JoinContext> context
                     joinNode.getDistributionType(),
                     joinNode.getDynamicFilters());
 
-            if (isTopJoin) {
-                return transformJoin(newJoinNode, joinContext);
+            if (isTopJoin && context.get().needsPayloadRejoin()) {
+                PlanNode payloadJoin = transformJoin(newJoinNode, joinContext);
+                // do a final check that the rewrite didn't lose any columns (can happen if there are intermediate projections on non-join keys that get hidden because of the DISTINCT keys computation
```

Comment:
```suggestion
                // do a final check that the rewrite didn't lose any columns (can happen if there are intermediate projections on non-join keys that get hidden because of the DISTINCT keys computation)
```